src/libraries/System.Formats.Tar/src/System/Formats/Tar/PosixTarEntry.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Formats.Tar
+{
+    /// <summary>
+    /// Abstract class that represents a tar entry from an archive of a format that is based on the POSIX IEEE P1003.1 standard from 1988. This includes the formats <see cref="TarFormat.Ustar"/> (represented by the <see cref="UstarTarEntry"/> class), <see cref="TarFormat.Pax"/> (represented by the <see cref="PaxTarEntry"/> class) and <see cref="TarFormat.Gnu"/> (represented by the <see cref="GnuTarEntry"/> class).
+    /// </summary>
+    /// <remarks>Formats that implement the POSIX IEEE P1003.1 standard from 1988, support the following header fields: <c>devmajor</c>, <c>devminor</c>, <c>gname</c> and <c>uname</c>.
+    /// Even though the <see cref="TarFormat.Gnu"/> format is not POSIX compatible, it implements and supports the Unix-specific fields that were defined in that POSIX standard.</remarks>
+    public abstract partial class PosixTarEntry : TarEntry
+    {
+        // Constructor used when reading an existing archive.
+        internal PosixTarEntry(TarHeader header, TarReader readerOfOrigin)
+            : base(header, readerOfOrigin)
+        {
+        }
+
+        // Constructor called when creating a new 'TarEntry*' instance that can be passed to a TarWriter.
+        internal PosixTarEntry(TarEntryType entryType, string entryName, TarFormat format)
+            : base(entryType, entryName, format)
+        {
+        }
+
+        /// <summary>
+        /// When the current entry represents a character device or a block device, the major number identifies the driver associated with the device.
+        /// </summary>
+        /// <remarks>Character and block devices are Unix-specific entry types.</remarks>
+        /// <exception cref="NotSupportedException">The entry does not represent a block device or a character device.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Cannot set a negative value.</exception>
+        public int DeviceMajor
+        {
+            get => _header._devMajor;
+            set
+            {
+                if (_header._typeFlag is not TarEntryType.BlockDevice and not TarEntryType.CharacterDevice)
+                {
+                    throw new NotSupportedException(SR.TarEntryBlockOrCharacterExpected);
```

Comment:
InvalidOperationException? or is this legal per spec, and we just don't implement it, but may do in future? we use NotSupportedException relatively rarely, mostly for things like methods we know are overridden.

Reply:
same for all your other NotSupportedExceptions. not sure what our pattern is.

Reply:
Ah didn't know that. I can change them all to InvalidOperationException.

These fields are properly implemented: They are only supported if the entry type is block or char device.

Reply:
Changed.